migration/handler.go

```diff
@@ -122,3 +122,77 @@ func (h *upgradeSchemaHandler) validate(ctx weave.Context, db weave.KVStore, tx
 
 	return &msg, nil
 }
+
+// SchemaRoutingHandler is a container that clubs toghether message handlers
+// for a single type message but different schema formats. Each handler is
+// registered together with the lowest schema version that it supports. For example
+//
+//   handler := SchemaRoutingHandler{
+//     1: &MyHandlerVersionAlpha{},
+//     7: &MyHandlerVersionBeta{},
+//   }
+//
+// In the above setup, messages with schema version 1 to 6 will be handled by
+// the alpha handler. Messages with schema version 7 and above are passed to
+// the beta handler.
+//
+// It is not allowed to use an empty SchemaRoutingHandler instance. It is not
+// allowed to register a handler for schema version zero.
+// All messages processed by this handler must implement Migratable interface.
+type SchemaRoutingHandler []weave.Handler
+
+var _ weave.Handler = (SchemaRoutingHandler)(nil)
+
+func (h SchemaRoutingHandler) Check(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.CheckResult, error) {
+	handler, err := h.selectHandler(tx)
+	if err != nil {
+		return nil, err
+	}
+	return handler.Check(ctx, db, tx)
+}
+
+func (h SchemaRoutingHandler) Deliver(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.DeliverResult, error) {
+	handler, err := h.selectHandler(tx)
+	if err != nil {
+		return nil, err
+	}
+	return handler.Deliver(ctx, db, tx)
+}
+
+// selectHandler returns the best fitting handler to process given transaction,
+// selected by introspecting the transaction message schema version.
+func (h SchemaRoutingHandler) selectHandler(tx weave.Tx) (weave.Handler, error) {
+	if len(h) == 0 {
```

Comment:
It would be interesting to check this in the constructor (eg. app init), not on first use of the handler.
So adding some `NewSchemaRoutingHandler(handlers []Handler)` method that validates may be useful. Or a `.Validate()` method to `SchemaRoutingHandler`. In any case, I would like mis-configuration to throw on app start.

Reply:
I agree. I did not provide it because I was not able to come up with a good way to declare it. Current declaration format is quite nice.
I will give it another shot.